src/cargo/util/toml/mod.rs

```diff
@@ -1006,9 +1012,14 @@ impl TomlManifest {
         let publish = match publish {
             Some(VecStringOrBool::VecString(ref vecstring)) => Some(vecstring.clone()),
             Some(VecStringOrBool::Bool(false)) => Some(vec![]),
-            None | Some(VecStringOrBool::Bool(true)) => None,
+            Some(VecStringOrBool::Bool(true)) => None,
+            None => version.is_none().then_some(vec![]),
         };
 
+        if version.is_none() && publish != Some(vec![]) {
+            bail!("`package.version` must be specified if `package.publish = true`");
```

Comment:
This error could be confusing if `publish = ['my-registry']`

Reply:
Hopefully the new message is a little better:

> `package.publish` requires `package.version` be specified